src/wayland/dmabuf/mod.rs

```diff
@@ -0,0 +1,465 @@
+//! To setup the dmabuf global, you will need to provide 2 things:
+//!
+//! - a list of the dmabuf formats you wish to support
+//! - an implementation of the `DmabufHandler` trait
+//!
+//! The list of supported format is just a `Vec<Format>`, where you will enter all the (format, modifier)
+//! couples you support.
+//!
+//! The implementation of the `DmabufHandler` trait will be called whenever a client has finished setting up
+//! a dma buffer. You will be handled the full details of the sclient's submission as a `BufferInfo` struct,
+//! and you need to validate it and maybe import it into your rendered. The `BufferData` associated type
+//! allows you to store any metadata of handle to the resource you need into the created `wl_buffer`,
+//! user data, to then retrieve it when it is attached to a surface to re-identify the dmabuf.
+//!
+//! ```
+//! # extern crate wayland_server;
+//! # extern crate smithay;
+//! use smithay::wayland::dmabuf::{DmabufHandler, BufferInfo, init_dmabuf_global};
+//!
+//! struct MyDmabufHandler;
+//!
+//! struct MyBufferData {
+//!     /* ... */
+//! }
+//!
+//! impl Drop for MyBufferData {
+//!     fn drop(&mut self) {
+//!         // This is called when all handles to this buffer have been dropped,
+//!         // both client-side and server side.
+//!         // You can now free the associated resource in your renderer.
+//!     }
+//! }
+//!
+//! impl DmabufHandler for MyDmabufHandler {
+//!     type BufferData = MyBufferData;
+//!     fn validate_dmabuf(&mut self, info: BufferInfo) -> Result<Self::BufferData, ()> {
+//!         /* validate the dmabuf and import it into your renderer state */
+//!         Ok(MyBufferData { /* ... */ })
+//!     }
+//! }
+//!
+//! // Once this is defined, you can in your setup initialize the dmabuf global:
+//!
+//! # fn main() {
+//! # let mut event_loop = wayland_server::calloop::EventLoop::<()>::new().unwrap();
+//! # let mut display = wayland_server::Display::new(event_loop.handle());
+//! // define your supported formats
+//! let formats = vec![
+//!     /* ... */
+//! ];
+//! let dmabuf_global = init_dmabuf_global(
+//!     &mut display,
+//!     formats,
+//!     MyDmabufHandler,
+//!     None // we don't provide a logger in this example
+//! );
+//! # }
+//! ```
+
+use std::{cell::RefCell, os::unix::io::RawFd, rc::Rc};
+
+pub use wayland_protocols::unstable::linux_dmabuf::v1::server::zwp_linux_buffer_params_v1::Flags;
+use wayland_protocols::unstable::linux_dmabuf::v1::server::{
+    zwp_linux_buffer_params_v1::{
+        Error as ParamError, RequestHandler as ParamRequestHandler, ZwpLinuxBufferParamsV1 as BufferParams,
+    },
+    zwp_linux_dmabuf_v1,
+};
+use wayland_server::{protocol::wl_buffer, Display, Global, NewResource};
+
+/// Representation of a Dmabuf format, as advertized to the client
+pub struct Format {
+    /// The format identifier.
+    ///
+    /// It must be a `DRM_FORMAT` code, as defined by the libdrm's drm_fourcc.h. The Linux kernel's DRM
```

Comment:
I think there is a typed variant of this in drm-rs. We probably want to use it. You cannot use dmabufs with the drm-system anyway. So dmabuf-support would need to depend on the `backend_drm` feature or pull in drm-rs at least.

Reply:
Oh, good point, do you mean https://docs.rs/drm/0.3.4/drm/buffer/enum.PixelFormat.html ?

Reply:
Yes, that should be it